src/models/chart-model/index.js

```diff
@@ -92,33 +103,100 @@ export default function createChartModel({
+    if (enableDataPagesExtract) {
+      // To avoid extract a chunk during progressive rendering
+      dataPages = layoutService.getDataPages();
+      layoutService.setDataPages([]);
+      enableDataPagesExtract = false;
+    }
+  }
+
+  function insertDataPages() {
+    layoutService.setDataPages(dataPages);
+    dataPages = [];
+    enableDataPagesExtract = true;
+  }
+
+  function updateMeta() {
+    const nodes = getPointNodesWithKey(chart, KEYS.COMPONENT.POINT);
+    meta.numVisibleBubbles = nodes.length;
+    meta.isLargeNumVisibleBubbles = nodes.length > layoutService.meta.largeNumDataPoints;
+  }
+
+  function renderOnce() {
+    cancelAnimationFrame(timer);
+    timer = requestAnimationFrame(() => {
       chart.update({
         partialData: true,
         excludeFromUpdate: EXCLUDE,
-        // transforms: [
-        //   {
-        //     key: KEYS.COMPONENT.POINT,
-        //     transform: { a: 1, b: 0, c: 0, d: 1, e: x, f: y },
-        //   },
-        // ],
       });
-      const nodes = getPointNodesWithKey(chart, KEYS.COMPONENT.POINT);
-      meta.numVisibleBubbles = nodes.length;
-      meta.isLargeNumVisibleBubbles = nodes.length > layoutService.meta.largeNumDataPoints;
-
-      // TODO: debounce -> interactionInProgess = false
-      // console.log('chart rendered in ', Date.now() - startTime, ' ms');
+      updateMeta();
     });
   }
 
+  const renderProgressive = () => {
+    const tempPages = layoutService.getDataPages();
+    if (!tempPages.length) return;
+    const dataSize = tempPages[0].qMatrix.length;
+    const nbrOfChunks = Math.ceil(dataSize / NUMBERS.CHUNK_SIZE);
+    if (nbrOfChunks <= 1) {
+      renderOnce();
+      return;
+    }
+
+    let renderCount = -1; // To clear before rendering
+
+    const renderChunk = () => {
+      cancelAnimationFrame(timer);
+      timer = requestAnimationFrame(() => {
+        extractDataPages();
+        const start = renderCount * NUMBERS.CHUNK_SIZE;
+        const end = (renderCount + 1) * NUMBERS.CHUNK_SIZE;
+        meta.progressive = renderCount === -1 ? false : { start, end };
+        const chunk = [
+          {
+            ...dataPages[0],
+            qMatrix: renderCount === -1 ? [] : dataPages[0].qMatrix.slice(start, end),
+          },
+        ];
+        layoutService.setDataPages(chunk);
+        chart.update({
+          partialData: true,
+          excludeFromUpdate:
+            renderCount === -1 || renderCount === nbrOfChunks - 1 ? EXCLUDE : EXCLUDE_DURING_PROGRESSIVE,
+        });
+        insertDataPages();
+        renderCount++;
+        if (renderCount < nbrOfChunks) {
+          renderChunk();
+        } else {
+          meta.progressive = false;
+          updateMeta();
+        }
+      });
+    };
+
+    renderChunk();
+  };
+
+  function updatePartial(interactionInProgress = false) {
+    meta.updateWithSettings = false;
+    trendLinesService.update();
+    if (interactionInProgress || !isProgressiveAllowed(layoutService)) {
```

Comment:
I think we should not pass in `interactionInProgress` to this function, but instead just do `viewHandler.getInteractionInProgress()`. Also, the `interactionInProgress` passed into `renderProgressive` is not being used anyway.

Reply:
We call `updatePartial` after fetching data. Can it be that `viewHandler.getInteractionInProgress() = true` when `updatePartial` is called after fetching data?
I think it is safer to pass a flag to `updatePartial` instead of checking `viewHandler.getInteractionInProgress()` inside `updatePartial`.
Note: Removed `interactionInProgress` from `renderProgressive`.